tools/boardgen.py

```diff
@@ -0,0 +1,517 @@
+            m = pin.enable_macro()
+            if m:
+                print("#if {}".format(m), file=out_header)
+
+            # #define pin_CPUNAME (...)
+            if cpu:
+                print(
+                    "#define pin_{:s} ({:s})".format(pin.name(), self._cpu_pin_pointer(pin)),
+                    file=out_header,
+                )
+
+            # #define pin_BOARDNAME (pin_CPUNAME)
+            if board:
+                for board_pin_name, _board_hidden in pin._board_pin_names:
+                    # Note: Hidden board pins are still available to C via the macro.
+                    # Note: The RHS isn't wrapped in (), which is necessary to make the
+                    # STATIC_AF_ macro work on STM32.
+                    print(
+                        "#define {:s}pin_{:s} pin_{:s}".format(
+                            self.board_name_define_prefix(),
+                            board_pin_name,
+                            pin.name(),
+                        ),
+                        file=out_header,
+                    )
+
+            if m:
+                print("#endif", file=out_header)
+
+    def print_pin_objects(self, out_source):
+        print(file=out_source)
+        for pin in self.available_pins():
+            m = pin.enable_macro()
+            if m:
+                print("#if {}".format(m), file=out_source)
+            print(
+                "{:s}machine_pin_obj_t pin_{:s}_obj = {:s};".format(
+                    "const " if pin.is_const() else "",
+                    pin.name(),
+                    pin.definition(),
+                ),
+                file=out_source,
+            )
+            if m:
+                print("#endif", file=out_source)
+
+    def print_pin_object_externs(self, out_header):
+        print(file=out_header)
+        for pin in self.available_pins():
+            m = pin.enable_macro()
+            if m:
+                print("#if {}".format(m), file=out_header)
+            print(
+                "extern {:s}machine_pin_obj_t pin_{:s}_obj;".format(
+                    "const " if pin.is_const() else "",
+                    pin.name(),
+                ),
+                file=out_header,
+            )
+            if m:
+                print("#endif", file=out_header)
+
+    def print_source(self, out_source):
+        self.print_pin_objects(out_source)
+        self.print_cpu_locals_dict(out_source)
+        self.print_board_locals_dict(out_source)
+
+    def print_header(self, out_header):
+        self.print_pin_object_externs(out_header)
+        self.print_defines(out_header)
+
+    # A port can override this if it has extra input files (e.g. af.csv) to load.
+    def load_inputs(self, out_source):
+        # Optionally load pins.csv to get cpu->board name mappings.
+        if self._enable_af and self.args.af_csv:
+            print("// --af-csv {:s}".format(self.args.af_csv), file=out_source)
+            self.parse_af_csv(self.args.af_csv)
```

Comment:
The original esp32 version of make-pins.py had the parsing of input files separated from the output printing.  First it parsed, then it printed.  But here those functions are mixed.

Don't you think it's cleaner to parse everything first, then print?  They can be separate functions/methods in this PinGenerator.

Reply:
> Don't you think it's cleaner to parse everything first, then print?

Honestly, not really? It's printing the information about the loading, not printing while parsing.

The alternative is more complicated -- keeping track of the filenames etc to print them later.

Even the original esp32 version did the prefix the same way it is here.

Reply:
OK, keep it as you have it.